OpenCV failed to run 9 tests in opencv_test_objdetect.exe

Hi All,

OpenCV failed to run 9 tests in opencv_test_objdetect.exe on Windows with VS2017 Update 7. Could you please help take a look at this? Thank you!

**You can repro this issue as the steps below:**
1. git clone https://github.com/opencv/opencv D:\OpenCV\src
2. git clone https://github.com/opencv/opencv_extra D:\OpenCV\src\extra
3. Open a VS 2017 x86 prompt and browse to D:\OpenCV
4. mkdir build_x86 && pushd build_x86
5. cmake -G "Visual Studio 15 2017" -DCMAKE_SYSTEM_VERSION=10.0.17134.0 -DWITH_IPP=ON -DBUILD_SHARED_LIBS=OFF -DBUILD_PERF_TESTS=ON -DBUILD_TESTS=ON -DBUILD_EXAMPLES=ON -DWITH_OPENCL=OFF -DBUILD_DOCS=OFF -DWITH_CUDA=OFF ..\src
6. msbuild /p:Configuration=Release;Platform=Win32 .\OpenCV.sln /t:Rebuild /m /p:BuildInParallel=true
7. .\bin\Release\opencv_test_objdetect.exe

**Error info:**
The whole log file please see attachment.
[opencv_test.log](https://github.com/opencv/opencv/files/2222804/opencv_test.log)
[----------] 9 tests from objdetect/Objdetect_QRCode
[ RUN      ] objdetect/Objdetect_QRCode.regression/0, where GetParam() = "20110817_030.jpg"
unknown file: error: C++ exception with description "OpenCV(3.4.2-dev) d:\opencv\src\modules\objdetect\src\qrcode.cpp:788: error: (-215:Assertion failed) !inarr.empty() in function 'cv::QRCodeDetector::detect'
" thrown in the test body.
[  FAILED  ] objdetect/Objdetect_QRCode.regression/0, where GetParam() = "20110817_030.jpg" (0 ms)

…

[ RUN      ] objdetect/Objdetect_QRCode.regression/8, where GetParam() = "qr_test_030.jpg"
unknown file: error: C++ exception with description "OpenCV(3.4.2-dev) d:\opencv\src\modules\objdetect\src\qrcode.cpp:788: error: (-215:Assertion failed) !inarr.empty() in function 'cv::QRCodeDetector::detect'
" thrown in the test body.
[  FAILED  ] objdetect/Objdetect_QRCode.regression/8, where GetParam() = "qr_test_030.jpg" (0 ms)
[----------] 9 tests from objdetect/Objdetect_QRCode (2 ms total)

--- a/modules/objdetect/test/test_qrcode.cpp
+++ b/modules/objdetect/test/test_qrcode.cpp
@@ -5,10 +5,9 @@
 #include "test_precomp.hpp"
 
 
-namespace opencv_test
-{
+namespace opencv_test { namespace {
 
-String qrcode_images_name[] = {
+std::string qrcode_images_name[] = {
     "20110817_030.jpg",
     "20110817_048.jpg",
     "img_20120226_161648.jpg",
@@ -25,24 +24,25 @@ String qrcode_images_name[] = {
 
 TEST(Objdetect_QRCode, generate_test_data)
 {
-    String root = cvtest::TS::ptr()->get_data_path() + "qrcode/";
-    String dataset_config = cvtest::TS::ptr()->get_data_path() + "qrcode/dataset_config.json";
+    const std::string root = "qrcode/";
+    const std::string dataset_config = findDataFile(root + "dataset_config.json");
     FileStorage file_config(dataset_config, FileStorage::WRITE);
 
     file_config << "test_images" << "[";
-    size_t images_count = sizeof(qrcode_images_name) / sizeof(String);
+    size_t images_count = sizeof(qrcode_images_name) / sizeof(qrcode_images_name[0]);
     for (size_t i = 0; i < images_count; i++)
     {
         file_config << "{:" << "image_name" << qrcode_images_name[i];
-        String image_path = root + qrcode_images_name[i];
-        std::vector<Point> transform;
+        std::string image_path = findDataFile(root + qrcode_images_name[i]);
+        std::vector<Point> corners;
         Mat src = imread(image_path, IMREAD_GRAYSCALE);
-        EXPECT_TRUE(detectQRCode(src, transform));
+        ASSERT_FALSE(src.empty()) << "Can't read image: " << image_path;
+        EXPECT_TRUE(detectQRCode(src, corners));
         file_config << "x" << "[:";
-        for (size_t j = 0; j < transform.size(); j++) { file_config << transform[j].x; }
+        for (size_t j = 0; j < corners.size(); j++) { file_config << corners[j].x; }
         file_config << "]";
         file_config << "y" << "[:";
-        for (size_t j = 0; j < transform.size(); j++) { file_config << transform[j].y; }
+        for (size_t j = 0; j < corners.size(); j++) { file_config << corners[j].y; }
         file_config << "]" << "}";
     }
     file_config << "]";
@@ -51,65 +51,65 @@ TEST(Objdetect_QRCode, generate_test_data)
 
 #else
 
-typedef testing::TestWithParam< String > Objdetect_QRCode;
+typedef testing::TestWithParam< std::string > Objdetect_QRCode;
 TEST_P(Objdetect_QRCode, regression)
 {
-    String root = cvtest::TS::ptr()->get_data_path() + "qrcode/";
-    String dataset_config = cvtest::TS::ptr()->get_data_path() + "qrcode/dataset_config.json";
-    FileStorage file_config(dataset_config, FileStorage::READ);
+    const std::string name_current_image = GetParam();
+    const std::string root = "qrcode/";
     const int pixels_error = 3;
 
-    std::vector<Point> corners;
-    String image_path = root + String(GetParam());
+    std::string image_path = findDataFile(root + name_current_image);
     Mat src = imread(image_path, IMREAD_GRAYSCALE);
+    ASSERT_FALSE(src.empty()) << "Can't read image: " << image_path;
+
+    std::vector<Point> corners;
     EXPECT_TRUE(detectQRCode(src, corners));
 
-    if (file_config.isOpened())
+    const std::string dataset_config = findDataFile(root + "dataset_config.json", false);
+    FileStorage file_config(dataset_config, FileStorage::READ);
+    ASSERT_TRUE(file_config.isOpened()) << "Can't read validation data: " << dataset_config;
     {
         FileNode images_list = file_config["test_images"];
-        int index = 0, images_count = static_cast<int>(images_list.size());
-        ASSERT_GT(images_count, 0);
+        size_t images_count = static_cast<size_t>(images_list.size());
+        ASSERT_GT(images_count, 0u) << "Can't find validation data entries in 'test_images': " << dataset_config;
 
-        bool runTestsFlag = false;
-        String name_current_image = String(GetParam());
-        for (; index < images_count; index++)
+        for (size_t index = 0; index < images_count; index++)
         {
-            String name_test_image = images_list[index]["image_name"];
+            FileNode config = images_list[(int)index];
+            std::string name_test_image = config["image_name"];
             if (name_test_image == name_current_image)
             {
                 for (int i = 0; i < 4; i++)
                 {
-                    int x = images_list[index]["x"][i];
-                    int y = images_list[index]["y"][i];
+                    int x = config["x"][i];
+                    int y = config["y"][i];
                     EXPECT_NEAR(x, corners[i].x, pixels_error);
                     EXPECT_NEAR(y, corners[i].y, pixels_error);
                 }
-                runTestsFlag = true;
+                return; // done
             }
         }
-        if (!runTestsFlag)
-        {
-            std::cout << "Not found results for " << name_current_image;
-            std::cout << " image in dataset_config.json file." << std::endl;
-        }
-
-        file_config.release();
-    }
-    else
-    {
-        std::cout << " Not found dataset_config.json file." << std::endl;
+        std::cerr
+            << "Not found results for '" << name_current_image
+            << "' image in config file:" << dataset_config << std::endl
+            << "Re-run tests with enabled UPDATE_QRCODE_TEST_DATA macro to update test data."
+            << std::endl;
     }
 }
 
-INSTANTIATE_TEST_CASE_P(objdetect, Objdetect_QRCode, testing::ValuesIn(qrcode_images_name));
+INSTANTIATE_TEST_CASE_P(/**/, Objdetect_QRCode, testing::ValuesIn(qrcode_images_name));
 
-TEST(Objdetect_QRCode, not_found_qrcode)
+
+
+TEST(Objdetect_QRCode_basic, not_found_qrcode)
 {
     std::vector<Point> corners;
     Mat zero_image = Mat::zeros(256, 256, CV_8UC1);
     EXPECT_FALSE(detectQRCode(zero_image, corners));
 }
 
-#endif
 
-} // namespace
+
+#endif // UPDATE_QRCODE_TEST_DATA
+
+}} // namespace

